File: Problem 016 - Power digit sum.cpp
// Find the sum of the digits of 2^1000
#include <iostream>
#include <vector>

#define POWER 1000

using namespace std;


int main()
{
    // Keep a vector to store the digits in since this number will be too big to store normally.
    vector<int> digits = {1};
    for (int i = 0; i < POWER; i++){
        // We start doubling our digits at the back of the vector (biggest digit first)
        for (int j = digits.size() - 1; j >= 0; j--){
            // We double the value, then check if it has become bigger than 10.
            int newval = digits.at(j) * 2;
            if (newval >= 10){
                // If it is bigger than 10, we remove the new digit and either:
                newval %= 10;
                // Add 1 to the next digit, or:
                if (j < digits.size() - 1){
                    digits.at(j+1) += 1;
                }
                // Create a new digit with value 1 if this was already the biggest.
                else {
                    digits.push_back(1);
                }
            }
            // Remember to store the new digit.
            digits.at(j) = newval;
        }
    }
    // Now sum up the digits.
    int sum = 0;
    for (int i = digits.size() - 1; i >= 0; i--){
        sum += digits.at(i);
    }
    cout<<"The sum of the digits of 2^"<<POWER<<" is: "<<sum;
    return 0;
}

File: LUA/Problem016 - Power digit sum.lua
--[[
2^15 = 32768 and the sum of its digits is 3 + 2 + 7 + 6 + 8 = 26.

What is the sum of the digits of the number 2^1000?
]]

-- For this solution I am not using a bignum library. I will attempt to achieve the correct result through use of tables

-- A limit we use to set the number of digits each individual part of our number can have. The bigger it is, the fewer multiplications we need to do.
limit = 10000000000000
-- Our target number of powers to get to
target = 1000

-- Create our table of numbers
numbers = {}

-- Initialize the first value to 1
numbers[1] = 1

for i = 1, target do
	-- Double our numbers from large to small
	for j = #numbers, 1, -1 do
		numbers[j] = numbers[j] * 2
		-- If a number gets too large for its container, we move the newly created largest digit to the next container
		if numbers[j] > limit then
			-- nil check. If it is not, simply add one to the next container. If it is, then we create a new entry with value 1 in our table
			if numbers[j+1] ~= nil then
				numbers[j+1] = numbers[j+1] + 1
			else
				numbers[j+1] = 1
			end
			-- And dont forget to remove the new digit from our old number
			numbers[j] = numbers[j] - limit
		end
	end
end

-- Now lets sum up our numbers. Since we store them multiple digits at a time we need to extract the individual digits from the numbers now
sum = 0
for i = #numbers, 1, -1 do
	while numbers[i] ~= 0 do
		sum = sum + (numbers[i]%10)
		numbers[i] = math.floor(numbers[i] / 10)
	end
end

-- Write the final answer
io.write(string.format("The sum of the digits of the number 2^%d is %d", target, sum))